File: java/src/basesort/ShellSort.java
package basesort;

import java.util.Arrays;

public class ShellSort {
    public static void shellSort(int nums[]){
        int length = nums.length;
        for(int gap = length>>1; gap > 0; gap >>= 1){
            for(int i = gap; i < length; i++){
                int temp = nums[i], j;
                for(j = i -gap; j >= 0 && nums[j] > temp; j -= gap){
                    nums[j+gap] = nums[j];
                }
                nums[j+gap] = temp;
            }
        }
    }

    public static void main(String[] args) {
        int[] testNums = {16, 3, 7, 1, 9, 8, 2, 4};
        shellSort(testNums);
        System.out.println(Arrays.toString(testNums));
    }
}

File: cpp/cpp/base_sort/ShellSort.cpp
#include <iostream>
#include "ShellSort.h"

using namespace std;

ShellSort::ShellSort(){}

void ShellSort::shellSort(int nums[], int length) {
	for (int gap = length >>2 ; gap > 0; gap >>= gap) {
		for (int i = gap; i < length; i++) {
			int temp = nums[i], j;
			for ( j = i - gap; j >= 0 && nums[j] > temp; j -= gap)
				nums[j + gap] = nums[j];
			nums[j + gap] = temp;
		}
	}
	for (int a = 0; a < length; a++) {
		cout << nums[a] << endl;
	}
}
/*
void main() {
	ShellSort sort;
	int input[7] = { 10, 21, 9, 7, 1, 42, 37 };
	int length = sizeof(input) / sizeof(input[0]);
	sort.shellSort(input, length);
}
*/
